up-l1/someip.adoc

```diff
@@ -24,164 +26,207 @@ SPDX-License-Identifier: Apache-2.0
+|===
+
+NOTES::
+* `ServiceID` and `MethodID` are defined in *SOME/IP Header format*, specified by AUTOSAR `[PRS_SOMEIP_00030]`.
+* `EventID` is an alias for `MethodID`, but used for events. `[PRS_SOMEIP_00245]`
+* `InstanceID` is defined in *SOME/IP-SD Service Entry Type*, specified by AUTOSAR `[PRS_SOMEIPSD_00268]`.
+* `InstanceID` is not present in *SOME/IP Header*, but is needed for _Service Discovery_ and events and *MUST* be provided for SOME/IP libraries (e.g. `vsomeip`). `[PRS_SOMEIP_00162]`
+* Additional mappings *MAY* be required by the underlying SOME/IP library, e.g. Service Major/Minor version. Such values *MAY* be pre-configured
 
-a| `ue_version_major` a| `Interface Version`
-a| `Interface Version` field of *SOME/IP Header*. `[PRS_SOMEIP_00053]`.
+=== UAttributes
 
-Maps to `Major Version` for the `Service ID`. If set, it *SHOULD* be passed to underlying *SOME/IP* library to override it's default _Service Major Version_
+The following sections highlight the mapping of link:../basics/uattributes.adoc[uAttributes] to *SOME/IP* fields.
 
-a| `resource_id` a| `Method ID`/`Event ID` | Identifier of the method/event placed in the
-lower 16 bits of the *SOME/IP Header* `Message ID`. `[PRS_SOMEIP_00034]`
+==== Common UMessage Attributes
 
+.Common UMessage Attributes
+[[common-umessage-attributes]]
+[cols="1m,4a", options="header,autowidth"]
 |===
+| UMessage | SOME/IP Mapping
+
+| id | UUid to SOME/IP `MessageID` as explained in <<uuid-mapping,UUIds>>.
 
-*NOTES:*
+| priority | Not available in SOME/IP.
 
-    * `Service ID` and `Method ID` are defined in *SOME/IP Header format*, specified by AUTOSAR `[PRS_SOMEIP_00030]`.
-    * `Event ID` is an alias for `Method ID`, but used for events. `[PRS_SOMEIP_00245]`
-    * `Instance ID` is defined in *SOME/IP-SD Service Entry Type*, specified by AUTOSAR `[PRS_SOMEIPSD_00268]`.
-    * `Instance ID` is not present in *SOME/IP Header*, but is needed for _Service Discovery_ and events and *MUST* be provided for *SOME/IP* libraries (e.g. `vsomeip`). `[PRS_SOMEIP_00162]`
-    * Additional mappings *MAY* be required by the underlying *SOME/IP* library, e.g. Service Major/Minor version. Such values *MAY* be pre-configured
+* A `NOTIFICATION` *MAY* use the default priority.
+* A `RESPONSE` *SHOULD* use the `REQUEST` priority.
 
+| ttl | Should be checked for incoming `UMESSAGE_TYPE_REQUEST` messages.
+
+* SOME/IP `REQUEST` *SHOULD* not be sent if incoming UMessage has `ttl > 0` and timestamp from `id` + `ttl` has already expired un *UE* host.
+* If present `ttl` *SHOULD* be used to limit the time a `Request` is kept in the cache for correlation with a `RESPONSE`.
+* If needed `ttl=0` *MAY* be used to indicate `REQUEST_NO_RETURN` (currently not allowed)
+
+NOTE: Unix timestamps between hosts are very likely to be different (even with `ntp`, especially for cloud events), so the transport *MAY* assume the local timestamp of receiving the request, especially for `ttl < 100ms`.
```

Comment:
About the NOTE?

I mean relying on UUID timestamp to be in sync on different hosts even in the vehicle network is a bit optimistic. Especially if requests have very small `ttl`, it may randomly expire them.

My idea addresses this issue: For local (in vehicle) uMessage requests with small enough `ttl` we may assume the time between generating the request UUID and receiving it in uE is almost identical. So in that edge case `ttl` should be used as response timeout. Consider the measured vsomeip request -> response is ~17ms.
This means in the ideal case tts < 20ms will either expire or not being send at all depending on sender and receiver clock deviation and network latency.

For cloud events it makes much more sense to check timestamp from UUID, but probably they should also use a bigger ttl as well.

So this is an optional solution the problem, if you want I'll remove it.

Reply:
Still do not understand. Are you saying we should not check ttl for in-vehicle use cases? We can't control app developers doing the wrong thing (putting a timeout of 1ms for something that should take 1000ms, etc...)

Reply:
It really depends if UMessage Request may be delayed in in-vehicle use cases.
If we assume just network latency for delivering the request, it simplifies timeout handling in some/ip bridge.
Then `ttl` can be used as RESPONSE timeout.

Otherwise `ttl` is used for both REQUEST validation and the remaining time as RESPONSE timeout.
Including the UUID timestamp variability between hosts it looks that small `ttl` values will cause undefined behavior, e.g:
  - sending the REQUEST, timing out for RESPONSE
  - sending the REQUEST and RESPONSE in time
  - not sending REQUEST at all

Reply:
The purpose of the ttl in the request really was for how long should the request be active for before we give up and declare a failure (and generate a `DEADLINE_EXCEEDED` error). So I think we are saying the same thing actually.